### DISCUSSION TITLE
Create an ARM target?

### DISCUSSION THREAD

**Discussion by pchapin** [open-watcom/open-watcom-v2#847] [Ideas]
In my day job I have been working with various ARM architecture processors. It sure seems like ARM has come to dominate the embedded systems market. What would be involved in creating an ARM target for Open Watcom? Clearly the code generator for the C compiler would need to be extended. Obviously there are other tools that would need, ideally, to be extended as well (the linker, librarian, and assembler come to mind... although it might be possible to use third party tools for some of these things at first). Then there is the matter of the C runtime library (but how much of that is architecture-specific?). Still, I wonder if it would be feasible to make something minimal that could target a bare board ARM device with a minimum of fuss. I'm probably being naïve. 

Thoughts?

**Comment by FredEckert** [comment]
I am currently working on a strategy to develop a single c code base for an embedded system that can be compiled for three targets: DOS4G/W, win32, and ARM. Right now, I'm just working with x86 using DOS4G/W and win32. DOS4G is for the target hardware and win32 is for a simulation of the target. For DOS4G, I build with OW, for win32, I build with VS2019, for ARM, I don't yet know what I would use... I agree that embedded systems are moving away from x86. An Open Watcom ARM target would be very interesting.

**Reply by jmalak** [reply]
I don't know how ARM architecture processors differ by instruction set.
I am not familiar with ARM any way.
If ARM architecture is like RISC then adoption of existing OW architecture (MIPS, PPC or AXP) could help.
My main task for OW Code Generator is to extend it for 64-bit Intel but it could be possible to do some work for ARM too.
You are right that ARM architecture is more and more used for embedded systems.
There is more questions about ARM infrastructure about ARM standards etc. which should be implemented by OW.

**Reply by pchapin** [reply]
ARM is RISC-like, although perhaps not "pure" RISC (whatever that means, exactly). I would love to work on a project like this, but then there are a lot of other projects that I would love to work on so... you know how that goes. If I get a little time maybe I'll check out those existing RISC code generators and see what I think.

You're right that there are a bunch of ARM standards that would also need to be thought about. Even if it were "easy" to hack one of the existing code generators to generate ARM machine code, that would just be the starting point. Still, it sounds fun!
